src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,144 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+from builtins import list, open, str
+from collections import defaultdict, namedtuple
+
+from future.moves.itertools import zip_longest
+from six import string_types
+
+from pants.base.build_environment import get_buildroot
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """A reporter to capture workunit data into a JSON structure."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report file goes.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+    # The main report, and associated tool outputs, go under this dir.
+    self._json_dir = settings.json_dir
+
+    self._buildroot = get_buildroot()
+    self._report_file = None
+
+    # Results of the state of the build.
+    self._results = {}
+    # We use a stack to track the nested workunit traversal of each root
+    self._root_id_to_workunit_stack = defaultdict(list)
+
+  def report_path(self):
+    """The path to the main report file."""
+
+    return os.path.join(self._json_dir, 'build.json')
+
+  def open(self):
+    """Implementation of Reporter callback."""
+
+    safe_mkdir(os.path.dirname(self._json_dir))
```

Comment:
Why are you taking the dirname here? You need to `mkdir` this `json_dir` itself, no?

Although per my comment above, this can be `os.path.dirname(self.report_path)` and you don't need `self._json_dir` at all.